d3dcore/include/d3dcore/d3d_info.h

```diff
@@ -0,0 +1,253 @@
+#pragma once
+
+#define INITGUID
+#include "dxdiag.h"
+#include <Windows.h>
+
+#include "d3d_utils.h"
+
+#include <string>
+#include <memory>
+#include <assert.h>
+#include <iostream>
+#include <string>
+#include <any>
+
+
+//-----------------------------------------------------------------------------
+// Defines, and constants
+//-----------------------------------------------------------------------------
+#define SAFE_DELETE(p)       { if(p) { delete (p);     (p)=NULL; } }
+#define SAFE_DELETE_ARRAY(p) { if(p) { delete[] (p);   (p)=NULL; } }
+#define SAFE_RELEASE(p)      { if(p) { (p)->Release(); (p)=NULL; } }
+#define SAFE_BSTR_FREE(x)    if(x) { SysFreeString( x ); x = NULL; }
+#define EXPAND(x)            x, sizeof(x)/sizeof(TCHAR)
+
+
+
+namespace d3dcore {
+
+	/*enum class SDKVersion {
+	DX9: DXDIAG_DX9_SDK_VERSION,
+	DX11: 112,		
+	};*/
+
+	template<typename Container>
+	class D3DInfo
+	{
+	public:
+		D3DInfo();
+		~D3DInfo();
+
+
+	private:
+		//property
+		IDxDiagProvider* m_dxdiag_provider;
+		IDxDiagContainer* m_dxdiag_root;
+		std::shared_ptr<Container> m_root_container;
+		//private method
+		std::string get_property_value(IDxDiagContainer* dxContainer, WCHAR* propName); //todo: update with sdt::any
+		
+		
+	public:
+		void init(int dxVersion = DXDIAG_DX9_SDK_VERSION);
+		void traverse(IDxDiagContainer*, std::shared_ptr<Container> parent);
+		void query_info();
+		std::shared_ptr<Container> get_root_container();
+	};
+
+}
+
+
+using namespace d3dcore;
+
+template<typename Container>
+D3DInfo<Container>::D3DInfo()
+{
+	m_dxdiag_provider = nullptr;
```

Comment:
Even if you have to define members in c-tor, standard C++ way is preferred
m_var{value}